base/reducedim.jl

```diff
@@ -143,19 +144,12 @@ let
         [AbstractArray{t} for t in uniontypes(BitIntFloat)]...,
         [AbstractArray{Complex{t}} for t in uniontypes(BitIntFloat)]...}
 
-    global reducedim_init(f::typeof(identity), op::typeof(+), A::T, region) =
-        reducedim_initarray(A, region, zero(eltype(A)))
-    global reducedim_init(f::typeof(identity), op::typeof(*), A::T, region) =
-        reducedim_initarray(A, region, one(eltype(A)))
-    global reducedim_init(f::Union{typeof(abs),typeof(abs2)}, op::typeof(+), A::T, region) =
-        reducedim_initarray(A, region, real(zero(eltype(A))))
-    global reducedim_init(f::Union{typeof(abs),typeof(abs2)}, op::typeof(*), A::T, region) =
-        reducedim_initarray(A, region, real(one(eltype(A))))
+    global reducedim_init(f, op::Union{typeof(+),typeof(add_sum)}, A::T, region) =
```

Comment:
I think it's not safe to generalize this to `f ∉ [identity, abs, abs2]` because we need the reduction operator to be type stable over whatever is produced by the `f` used for `mapreduce`. Nevertheless, it's a nice simplification to have, but I think we will need to continue hardcoding the list of acceptable `map` functions.

Reply:
Won't `mapreduce_single(f, op, zero(eltype(A)))` should handle that?

Reply:
In general, for some choices of `f`, this assumes that `mr_single(f, +, 0) + f(0)`, etc., will be of the same concrete type of `mr_single(f, +, 0)`, which one can design `f` to make it fail. This seems to be where `_reducedim_init` is more general and guards against. Of course it is also possible to thwart `_reducedim_init` with badly behaving `f`, so 😕.

Reply:
Although I'm not actually sure `_reducedim_init` does what it should in these type unstable cases, so maybe it's best to leave cleanup of `_reducedim_init` to later.